src/Microsoft.SqlTools.ServiceLayer/EditData/EditDataService.cs

```diff
@@ -0,0 +1,328 @@
+        }
+
+        #endregion
+
+        #region Member Variables 
+
+        private readonly ConnectionService connectionService;
+
+        private readonly IEditMetadataFactory metadataFactory;
+
+        private readonly QueryExecutionService queryExecutionService;
+
+        private readonly Lazy<ConcurrentDictionary<string, Session>> editSessions = new Lazy<ConcurrentDictionary<string, Session>>(
+            () => new ConcurrentDictionary<string, Session>());
+
+        #endregion
+
+        #region Properties
+
+        /// <summary>
+        /// Dictionary mapping OwnerURIs to active sessions
+        /// </summary>
+        internal ConcurrentDictionary<string, Session> ActiveSessions => editSessions.Value;
+
+        #endregion
+
+        /// <summary>
+        /// Initializes the edit data service with the service host
+        /// </summary>
+        /// <param name="serviceHost">The service host to register commands/events with</param>
+        public void InitializeService(ServiceHost serviceHost)
+        {
+            // Register handlers for requests
+            serviceHost.SetRequestHandler(EditCreateRowRequest.Type, HandleCreateRowRequest);
+            serviceHost.SetRequestHandler(EditDeleteRowRequest.Type, HandleDeleteRowRequest);
+            serviceHost.SetRequestHandler(EditDisposeRequest.Type, HandleDisposeRequest);
+            serviceHost.SetRequestHandler(EditInitializeRequest.Type, HandleInitializeRequest);
+            serviceHost.SetRequestHandler(EditRevertRowRequest.Type, HandleRevertRowRequest);
+            serviceHost.SetRequestHandler(EditUpdateCellRequest.Type, HandleUpdateCellRequest);
+
+            // Register handler for shutdown event
+            serviceHost.RegisterShutdownTask((shutdownParams, requestContext) =>
+            {
+                Dispose();
+                return Task.FromResult(0);
+            });
+        }
+
+        #region Request Handlers
+
+        internal async Task HandleCreateRowRequest(EditCreateRowParams createParams,
+            RequestContext<EditCreateRowResult> requestContext)
+        {
+            try
+            {
+                Session session = GetActiveSessionOrThrow(createParams.OwnerUri);
+
+                // Create the row and get send the ID of the row back
+                long newRowId = session.CreateRow();
+                EditCreateRowResult createResult = new EditCreateRowResult
+                {
+                    NewRowId = newRowId
+                };
+                await requestContext.SendResult(createResult);
+            }
+            catch (Exception e)
+            {
+                await requestContext.SendError(e.Message);
+            }
+        }
+
+        internal async Task HandleDeleteRowRequest(EditDeleteRowParams deleteParams,
+            RequestContext<EditDeleteRowResult> requestContext)
+        {
+            try
+            {
+                Session session = GetActiveSessionOrThrow(deleteParams.OwnerUri);
```

Comment:
All the session-dependent methods could be done from a single function - HandleSessionRequest<T>. We have a general pattern for this (see the TSqlFormatterService) as shown below. This would be a specialized version, that also takes in the session - something like below

```
        private async Task HandleRequest<T>(Func<Session,Task<T>> handler, RequestContext<T> requestContext, string requestType, Uri ownerUri)
        {
            Logger.Write(LogLevel.Verbose, requestType);

            try
            {
                Session session = GetActiveSessionOrThrow(ownerUri);
                T result = await handler(session);
                await requestContext.SendResult(result);
            }
            catch (Exception ex)
            {
                await requestContext.SendError(ex.ToString());
            }
        }
```